hplp example with S3 bucket

SF-N opened the pull request.

juntyr commented on 02-data-sources/README.md:
```diff
@@ -9,6 +9,7 @@ The **Online Laboratory for Data Compression** is built so you can explore vario
 - [`02-remote.ipynb`](02-remote.ipynb): open large remote datasets using `fsspec`, `kerchunk`, and `zarr`
 - [`03-cdsapi.ipynb`](03-cdsapi.ipynb): download small datasets from the Climate Data Store using the `cdsapi`
 - [`04-ecmwfapi.ipynb`](04-ecmwfapi.ipynb): download small datasets from the ECMWF Archive using the `ecmwfapi`
+- [`05-hplp-s3bucket.ipynb`](05-hplp-s3bucket.ipynb): open hplp-experiment data from the ECMWF S3 bucket
```
```suggestion
- [`05-hplp-s3bucket.ipynb`](05-hplp-s3bucket.ipynb): open hplp-experiment datasets from the ECMWF S3 bucket
```

juntyr replied:
Is the data large?

SF-N replied:
It is 175 GB in total.

juntyr commented on README.md:
```diff
@@ -46,6 +46,7 @@ The following is an overview of all notebooks:
   - [`02-remote.ipynb`](02-data-sources/02-remote.ipynb): open large remote datasets using `fsspec`, `kerchunk`, and `zarr`
   - [`03-cdsapi.ipynb`](02-data-sources/03-cdsapi.ipynb): download small datasets from the Climate Data Store using the `cdsapi`
   - [`04-ecmwfapi.ipynb`](02-data-sources/04-ecmwfapi.ipynb): download small datasets from the ECMWF Archive using the `ecmwfapi`
+  - [`05-hplp-s3bucket.ipynb`](02-data-sources/05-hplp-s3bucket.ipynb): open hplp-experiment data from the ECMWF S3 bucket
```
```suggestion
  - [`05-hplp-s3bucket.ipynb`](02-data-sources/05-hplp-s3bucket.ipynb): open hplp-experiment datasets from the ECMWF S3 bucket
```

juntyr replied:
Same as above, is the data large?

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
The notebook title could be `Example: Loading GRIB files from the hplp experiment from an S3 Bucket`

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
The notebook is missing the caution notes that the other notebooks have

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
Would it be possible to have a hplp experiment link that doesn't require logging in?

SF-N replied:
Yes, the new one should work.

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
"instead, they are accessed remotely via gribscan and corresponding.ref files, also stored on the S3 bucket." -> "instead they are accessed remotely using `gribscan`, `fsspec`, and `zarr` via their corresponding `.ref` metadata files (also hosted in the S3 bucket), which were produced up-front using `gribscan`."

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
When listing the parameters, could you also give their names? When giving model levels, could you give a brief comment on the heights that they map to? Perhaps something similar for the timesteps (how many hours do they correspond to)?

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
Please sort imports alphabetically per block, with a first block for system packages and a second block for third-party packages

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
Instead of manually downloading the JSON `.ref` file, you can use the following:

```python
xr.open_dataset(f"reference::{url}.ref", engine="zarr", consolidated=False)
```

SF-N replied:
Locally that works fine, but in the online lab, I get:
```
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
Cell In[35], line 1
----> 1 ds_pl = load_grib_files(leveltype="pl", gridtype="reduced_gg")
      2 ds_pl

Cell In[34], line 46, in load_grib_files(leveltype, gridtype, step, remap)
     44 print(f"Loading dataset {url}")
     45 storage_options = {"fo": fsspec.get_mapper(f"{url}.ref")} 
---> 46 return xr.open_dataset(f"reference::{url}.ref", engine="zarr", consolidated=False)

File /lib/python3.12/site-packages/xarray/backends/api.py:671, in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, inline_array, chunked_array_type, from_array_kwargs, backend_kwargs, **kwargs)
    659 decoders = _resolve_decoders_kwargs(
    660     decode_cf,
    661     open_backend_dataset_parameters=backend.open_dataset_parameters,
   (...)
    667     decode_coords=decode_coords,
    668 )
    670 overwrite_encoded_chunks = kwargs.pop("overwrite_encoded_chunks", None)
--> 671 backend_ds = backend.open_dataset(
    672     filename_or_obj,
    673     drop_variables=drop_variables,
    674     **decoders,
    675     **kwargs,
    676 )
    677 ds = _dataset_from_backend_dataset(
    678     backend_ds,
    679     filename_or_obj,
   (...)
    689     **kwargs,
    690 )
    691 return ds

File /lib/python3.12/site-packages/xarray/backends/zarr.py:1410, in ZarrBackendEntrypoint.open_dataset(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta, group, mode, synchronizer, consolidated, chunk_store, storage_options, stacklevel, zarr_version, zarr_format, store, engine, use_zarr_fill_value_as_mask)
   1408 filename_or_obj = _normalize_path(filename_or_obj)
   1409 if not store:
-> 1410     store = ZarrStore.open_group(
   1411         filename_or_obj,
   1412         group=group,
   1413         mode=mode,
   1414         synchronizer=synchronizer,
   1415         consolidated=consolidated,
   1416         consolidate_on_close=False,
   1417         chunk_store=chunk_store,
   1418         storage_options=storage_options,
   1419         stacklevel=stacklevel + 1,
   1420         zarr_version=zarr_version,
   1421         use_zarr_fill_value_as_mask=None,
   1422         zarr_format=zarr_format,
   1423     )
   1425 store_entrypoint = StoreBackendEntrypoint()
   1426 with close_on_error(store):

File /lib/python3.12/site-packages/xarray/backends/zarr.py:636, in ZarrStore.open_group(cls, store, mode, synchronizer, group, consolidated, consolidate_on_close, chunk_store, storage_options, append_dim, write_region, safe_chunks, stacklevel, zarr_version, zarr_format, use_zarr_fill_value_as_mask, write_empty)
    611 @classmethod
    612 def open_group(
    613     cls,
   (...)
    629     write_empty: bool | None = None,
    630 ):
    631     (
    632         zarr_group,
    633         consolidate_on_close,
    634         close_store_on_close,
    635         use_zarr_fill_value_as_mask,
--> 636     ) = _get_open_params(
    637         store=store,
    638         mode=mode,
    639         synchronizer=synchronizer,
    640         group=group,
    641         consolidated=consolidated,
    642         consolidate_on_close=consolidate_on_close,
    643         chunk_store=chunk_store,
    644         storage_options=storage_options,
    645         stacklevel=stacklevel,
    646         zarr_version=zarr_version,
    647         use_zarr_fill_value_as_mask=use_zarr_fill_value_as_mask,
    648         zarr_format=zarr_format,
    649     )
    651     return cls(
    652         zarr_group,
    653         mode,
   (...)
    660         use_zarr_fill_value_as_mask,
    661     )

File /lib/python3.12/site-packages/xarray/backends/zarr.py:1644, in _get_open_params(store, mode, synchronizer, group, consolidated, consolidate_on_close, chunk_store, storage_options, stacklevel, zarr_version, use_zarr_fill_value_as_mask, zarr_format)
   1640     if _zarr_v3():
   1641         # we have determined that we don't want to use consolidated metadata
   1642         # so we set that to False to avoid trying to read it
   1643         open_kwargs["use_consolidated"] = False
-> 1644     zarr_group = zarr.open_group(store, **open_kwargs)
   1645 close_store_on_close = zarr_group.store is not store
   1647 # we use this to determine how to handle fill_value

File /lib/python3.12/site-packages/zarr/hierarchy.py:1554, in open_group(store, mode, cache_attrs, synchronizer, path, chunk_store, storage_options, zarr_version, meta_array)
   1503 """Open a group using file-mode-like semantics.
   1504 
   1505 Parameters
   (...)
   1550 
   1551 """
   1553 # handle polymorphic store arg
-> 1554 store = _normalize_store_arg(
   1555     store, storage_options=storage_options, mode=mode, zarr_version=zarr_version
   1556 )
   1557 if zarr_version is None:
   1558     zarr_version = getattr(store, "_store_version", DEFAULT_ZARR_VERSION)

File /lib/python3.12/site-packages/zarr/hierarchy.py:1396, in _normalize_store_arg(store, storage_options, mode, zarr_version)
   1394 if store is None:
   1395     return MemoryStore() if zarr_version == 2 else MemoryStoreV3()
-> 1396 return normalize_store_arg(
   1397     store, storage_options=storage_options, mode=mode, zarr_version=zarr_version
   1398 )

File /lib/python3.12/site-packages/zarr/storage.py:201, in normalize_store_arg(store, storage_options, mode, zarr_version)
    199 else:
    200     raise ValueError("zarr_version must be either 2 or 3")
--> 201 return normalize_store(store, storage_options, mode)

File /lib/python3.12/site-packages/zarr/storage.py:171, in _normalize_store_arg_v2(store, storage_options, mode)
    169 if isinstance(store, str):
    170     if "://" in store or "::" in store:
--> 171         return FSStore(store, mode=mode, **(storage_options or {}))
    172     elif storage_options:
    173         raise ValueError("storage_options passed with non-fsspec path")

File /lib/python3.12/site-packages/zarr/storage.py:1384, in FSStore.__init__(self, url, normalize_keys, key_separator, mode, exceptions, dimension_separator, fs, check, create, missing_exceptions, **storage_options)
   1382 if protocol in (None, "file") and not storage_options.get("auto_mkdir"):
   1383     storage_options["auto_mkdir"] = True
-> 1384 self.map = fsspec.get_mapper(url, **{**mapper_options, **storage_options})
   1385 self.fs = self.map.fs  # for direct operations
   1386 self.path = self.fs._strip_protocol(url)

File /lib/python3.12/site-packages/fsspec/mapping.py:249, in get_mapper(url, check, create, missing_exceptions, alternate_root, **kwargs)
    218 """Create key-value interface for given URL and options
    219 
    220 The URL will be of the form "protocol://location" and point to the root
   (...)
    246 ``FSMap`` instance, the dict-like key-value store.
    247 """
    248 # Removing protocol here - could defer to each open() on the backend
--> 249 fs, urlpath = url_to_fs(url, **kwargs)
    250 root = alternate_root if alternate_root is not None else urlpath
    251 return FSMap(root, fs, check, create, missing_exceptions=missing_exceptions)

File /lib/python3.12/site-packages/fsspec/core.py:411, in url_to_fs(url, **kwargs)
    409     inkwargs["fo"] = urls
    410 urlpath, protocol, _ = chain[0]
--> 411 fs = filesystem(protocol, **inkwargs)
    412 return fs, urlpath

File /lib/python3.12/site-packages/fsspec/registry.py:303, in filesystem(protocol, **storage_options)
    296     warnings.warn(
    297         "The 'arrow_hdfs' protocol has been deprecated and will be "
    298         "removed in the future. Specify it as 'hdfs'.",
    299         DeprecationWarning,
    300     )
    302 cls = get_filesystem_class(protocol)
--> 303 return cls(**storage_options)

File /lib/python3.12/site-packages/fsspec/spec.py:81, in _Cached.__call__(cls, *args, **kwargs)
     79     return cls._cache[token]
     80 else:
---> 81     obj = super().__call__(*args, **kwargs)
     82     # Setting _fs_token here causes some static linters to complain.
     83     obj._fs_token_ = token

File /lib/python3.12/site-packages/fsspec/implementations/reference.py:697, in ReferenceFileSystem.__init__(self, fo, target, ref_storage_args, target_protocol, target_options, remote_protocol, remote_options, fs, template_overrides, simple_templates, max_gap, max_block, cache_size, **kwargs)
    695     with fsspec.open(fo, "rb", **dic) as f:
    696         logger.info("Read reference from URL %s", fo)
--> 697         text = json.load(f)
    698     self._process_references(text, template_overrides)
    699 else:
    700     # Lazy parquet refs

File /lib/python312.zip/json/__init__.py:293, in load(fp, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    274 def load(fp, *, cls=None, object_hook=None, parse_float=None,
    275         parse_int=None, parse_constant=None, object_pairs_hook=None, **kw):
    276     """Deserialize ``fp`` (a ``.read()``-supporting file-like object containing
    277     a JSON document) to a Python object.
    278 
   (...)
    291     kwarg; otherwise ``JSONDecoder`` is used.
    292     """
--> 293     return loads(fp.read(),
    294         cls=cls, object_hook=object_hook,
    295         parse_float=parse_float, parse_int=parse_int,
    296         parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)

File /lib/python312.zip/json/__init__.py:346, in loads(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    341     s = s.decode(detect_encoding(s), 'surrogatepass')
    343 if (cls is None and object_hook is None and
    344         parse_int is None and parse_float is None and
    345         parse_constant is None and object_pairs_hook is None and not kw):
--> 346     return _default_decoder.decode(s)
    347 if cls is None:
    348     cls = JSONDecoder

File /lib/python312.zip/json/decoder.py:337, in JSONDecoder.decode(self, s, _w)
    332 def decode(self, s, _w=WHITESPACE.match):
    333     """Return the Python representation of ``s`` (a ``str`` instance
    334     containing a JSON document).
    335 
    336     """
--> 337     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    338     end = _w(s, end).end()
    339     if end != len(s):

File /lib/python312.zip/json/decoder.py:353, in JSONDecoder.raw_decode(self, s, idx)
    344 """Decode a JSON document from ``s`` (a ``str`` beginning with
    345 a JSON document) and return a 2-tuple of the Python
    346 representation and the index in ``s`` where the document ended.
   (...)
    350 
    351 """
    352 try:
--> 353     obj, end = self.scan_once(s, idx)
    354 except StopIteration as err:
    355     raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
You can remove blocks [6] and [7] since those fixes have been merged into `earthkit.plots` already

juntyr commented on 02-data-sources/05-hplp-s3bucket.ipynb:
For the temperature plot, can you pass `units="degC"` to the plotting so we get the plot in degrees C?

SF-N replied:
I tried that, but it still plots (at least the color bar) in K. Please let me know if you have any suggestions about this.

juntyr reviewed (changes requested):
Thank you @SF-N for working on this notebook! It looks fantastic!

I left a bunch of comments, mostly minor nits, that should be quick to resolve (unfortunately Github only let me give the comments on the file itself and not the specific lines).

SF-N commented:
Thanks a lot for the review. I addressed all comments, the only ones that are still open are plotting in degrees C and opening the dataset without manually downloading the ref.

juntyr commented:
> Thanks a lot for the review. I addressed all comments, the only ones that are still open are plotting in degrees C and opening the dataset without manually downloading the ref.

I have no idea what's going on with the degC there ...

I think I observed the ref error before but found a way to fix it, I just need to remember what the fix was ;)

juntyr commented:
I pushed some minor cleanup tweaks.

I'll look at the two remaining issues later today (I hope), otherwise this is good to go and I'll merge it soon

SF-N commented:
Thanks a lot!

juntyr commented:
The ref loading seems to be an issue with how the ECMWF S3 bucket is configured (since it works on Finland's CSC bucket, as can be seen in the remote notebook). In particular, it seems that a HEAD request (or a GET request without specifying an explicit range header) to the URLs doesn't report the content-length or content-range. fsspec then gets confused and fails to load the length.

Interestingly enough, this issue doesn't happen with the GRIB files themselves but is perhaps triggered by the `.ref` extension?

You can technically work around it by requesting that fsspec not try to estimate the data size using:
```python
xr.open_dataset(
    f"reference::{url}.ref",
    engine="zarr",
    backend_kwargs={"storage_options": {"ref_storage_args": {"block_size": 0}}},
    consolidated=False,
)
```
but by that point we lose the simplicity of the URL.

For now, we can keep the code as-is (and improve it later if the S3 bucket behaviour changes)

SF-N commented:
> The ref loading seems to be an issue with how the ECMWF S3 bucket is configured (since it works on Finland's CSC bucket, as can be seen in the remote notebook). In particular, it seems that a HEAD request (or a GET request without specifying an explicit range header) to the URLs doesn't report the content-length or content-range. fsspec then gets confused and fails to load the length.
> 
> Interestingly enough, this issue doesn't happen with the GRIB files themselves but is perhaps triggered by the `.ref` extension?
> 
> You can technically work around it by requesting that fsspec not try to estimate the data size using:
> 
> ```python
> xr.open_dataset(
>     f"reference::{url}.ref",
>     engine="zarr",
>     backend_kwargs={"storage_options": {"ref_storage_args": {"block_size": 0}}},
>     consolidated=False,
> )
> ```
> 
> but by that point we lose the simplicity of the URL.
> 
> For now, we can keep the code as-is (and improve it later if the S3 bucket behaviour changes)

Thanks for investigating and the explanation. I also think we should keep the code as-is for now.

juntyr commented:
I have absolutely no idea about the earthkit units issue and why it works for other data. Perhaps we can open an issue for that?

juntyr reviewed (approved):
LGTM!
